### PR TITLE
[Model] Add `dgl.nn.CuGraphGATConv` model

### PR DESCRIPTION
**Pull request by tingyu66**
## Description

This PR adds a GATConv model Add `dgl.nn.CuGraphGATConv` that uses the accelerated sparse aggregation primitives in cugraph-ops. It requires `pylibcugraphops >= 23.02`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- [x] new nn.Module `CuGraphGATConv`
- [x] test to verify `CuGraphGATConv`

### Notes
Fixes rapidsai/cugraph-ops#181.

### CODE REVIEW

**Comment by tingyu66** [comment]
@mufeili This PR for CuGraphGATConv is also ready for review.

Compared to `GATConv`, the cugraph variant lacks support for:
- `attn_drop` option to drop out attention weights
- different feature sizes for source and destination nodes

Performance-wise, in the `examples/pytorch/gat/train.py` (full-graph) example with `pubmed` dataset, `CuGraphGATConv` takes 4e-03s per epoch, while GATConv takes 1.2e-02s. Runs with other datasets demonstrated similar speedup numbers.

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    from pylibcugraphops import make_fg_csr, make_mfg_csr
+    from pylibcugraphops.torch.autograd import mha_gat_n2n as GATConvAgg
+except ImportError:
+    has_pylibcugraphops = False
+else:
+    has_pylibcugraphops = True
+
+
+class CuGraphGATConv(nn.Module):
+    r"""Graph attention layer from `Graph Attention Network
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_gatconv.py:18]:
```suggestion
    r"""Graph attention layer from `Graph Attention Networks
```

**Comment by mufeili** [comment]
> @mufeili This PR for CuGraphGATConv is also ready for review.
> 
> Compared to `GATConv`, the cugraph variant lacks support for:
> 
> * `attn_drop` option to drop out attention weights
> * different feature sizes for source and destination nodes
> 
> Performance-wise, in the `examples/pytorch/gat/train.py` (full-graph) example with `pubmed` dataset, `CuGraphGATConv` takes 4e-03s per epoch, while GATConv takes 1.2e-02s. Runs with other datasets demonstrated similar speedup numbers.

How about model accuracy?

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        Dropout rate on feature. Defaults: ``0``.
+    negative_slope : float, optional
+        LeakyReLU angle of negative slope. Defaults: ``0.2``.
+    residual : bool, optional
+        If True, use residual connection. Defaults: ``False``.
+    activation : callable activation function/layer or None, optional.
+        If not None, applies an activation function to the updated node features.
+        Default: ``None``.
+    bias : bool, optional
+        If True, learns a bias term. Defaults: ``True``.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_gatconv.py:48]: What if the graph has zero-degree nodes?

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    def reset_parameters(self):
+        r"""Reinitialize learnable parameters."""
+
+        gain = nn.init.calculate_gain("relu")
+        nn.init.xavier_normal_(self.fc.weight, gain=gain)
+        nn.init.xavier_normal_(
+            self.attn_weights.view(2, self.num_heads, self.out_feats), gain=gain
+        )
+        if self.bias is not None:
+            nn.init.zeros_(self.bias)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_gatconv.py:138]: Missing `self.res_fc`

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    def forward(self, g, feat, max_in_degree=None):
+        r"""Forward computation.
+
+        Parameters
+        ----------
+        graph : DGLGraph
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_gatconv.py:145]:
```suggestion
        g : DGLGraph
```

**Review by mufeili — CHANGES_REQUESTED** [review]
done a pass

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py (hunk 3 above)
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_gatconv.py:138]: fixed.

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py (hunk 2 above)
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_gatconv.py:48]: Their output would be zero.

**Comment by tingyu66** [comment]
> > @mufeili This PR for CuGraphGATConv is also ready for review.
> > Compared to `GATConv`, the cugraph variant lacks support for:
> > 
> > * `attn_drop` option to drop out attention weights
> > * different feature sizes for source and destination nodes
> > 
> > Performance-wise, in the `examples/pytorch/gat/train.py` (full-graph) example with `pubmed` dataset, `CuGraphGATConv` takes 4e-03s per epoch, while GATConv takes 1.2e-02s. Runs with other datasets demonstrated similar speedup numbers.
> 
> How about model accuracy?

We get the same accuracy number as the [DGL](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/README.md) in those examples.

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/conv/cugraph_gatconv.py:48]: Is this consistent with the behavior of `GATConv`?

**Review by mufeili — APPROVED** [review]
I'm good.

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/conv/cugraph_gatconv.py (hunk 2 above)
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_gatconv.py:48]: Yes, `GATConv` also outputs zeros for zero-degree nodes.
